File: sdk/typescript/src/api/orchestration.ts
import type { HttpClient, RequestOptions } from "../http.js";
import type { WorkspaceProfile } from "./medulla.js";

export interface OrchestrationEvent {
  seq: number;
  role: "user" | "assistant" | "system";
  sender: string;
  body: string;
  ts: number;
  kind: string;
}

export interface IngestOrchestrationEventBody {
  protocol: number;
  counterpartAgentId: string;
  sessionId: string;
  event: OrchestrationEvent;
}

export interface OrchestrationWorldDiffQuery {
  session: string;
}

export interface OrchestrationMessagesQuery {
  after?: number;
}

export interface OrchestrationTool {
  name: string;
  description: string;
  parameters?: Record<string, unknown>;
}

export interface RunOrchestrationBody {
  input: string;
  sessionId?: string;
  flavor?: "tui" | "openhuman";
  tools?: OrchestrationTool[];
  options?: {
    workspaceProfiles?: WorkspaceProfile[];
    [key: string]: unknown;
  };
}

export interface OrchestrationToolResult {
  id: string;
  ok: boolean;
  result?: unknown;
  error?: string;
}

export interface ContinueOrchestrationRunBody {
  cycleId: string;
  toolResults?: OrchestrationToolResult[];
}

export class OrchestrationApi {
  constructor(private readonly http: HttpClient) {}

  ingestEvent<T = unknown>(
    body: IngestOrchestrationEventBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/orchestration/v1/events", body, options);
  }

  postWorldDiff<T = unknown>(body?: unknown, options?: RequestOptions): Promise<T> {
    return this.http.post<T>("/orchestration/v1/world-diff", body, options);
  }

  getWorldDiff<T = unknown>(
    query: OrchestrationWorldDiffQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>("/orchestration/v1/world-diff", {
      ...options,
      query: { ...query, ...options?.query },
    });
  }

  listSessions<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/orchestration/v1/sessions", options);
  }

  listSessionMessages<T = unknown>(
    id: string,
    query?: OrchestrationMessagesQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>(
      `/orchestration/v1/sessions/${encodeURIComponent(id)}/messages`,
      { ...options, query: { ...query, ...options?.query } },
    );
  }

  getSessionState<T = unknown>(
    id: string,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>(
      `/orchestration/v1/sessions/${encodeURIComponent(id)}/state`,
      options,
    );
  }

  run<T = unknown>(
    body: RunOrchestrationBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/orchestration/v1/run", body, options);
  }

  continueRun<T = unknown>(
    body: ContinueOrchestrationRunBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/orchestration/v1/run/continue", body, options);
  }
}

File: sdk/python/src/tinyhumans/api/orchestration.py
from __future__ import annotations

from typing import Any

from ..http import Json
from ..types import OrchestrationContinue, OrchestrationEventIngest, OrchestrationRun
from ._base import ApiNamespace, enc

__all__ = ["OrchestrationApi"]


class OrchestrationApi(ApiNamespace):
    """Device counterpart events, world state, and orchestration cycles."""

    def ingest_event(self, body: OrchestrationEventIngest, **kwargs: Any) -> Json:
        return self._http.post("/orchestration/v1/events", body=body, **kwargs)

    def update_world_diff(self, body: dict[str, Any], **kwargs: Any) -> Json:
        return self._http.post("/orchestration/v1/world-diff", body=body, **kwargs)

    def get_world_diff(self, session: str, **kwargs: Any) -> Json:
        return self._http.get(
            "/orchestration/v1/world-diff", query={"session": session}, **kwargs
        )

    def list_sessions(self, **kwargs: Any) -> Json:
        return self._http.get("/orchestration/v1/sessions", **kwargs)

    def list_messages(
        self, id: str, *, after: int | None = None, **kwargs: Any
    ) -> Json:
        return self._http.get(
            f"/orchestration/v1/sessions/{enc(id)}/messages",
            query={"after": after} if after is not None else None,
            **kwargs,
        )

    def get_session_state(self, id: str, **kwargs: Any) -> Json:
        return self._http.get(f"/orchestration/v1/sessions/{enc(id)}/state", **kwargs)

    def run(self, body: OrchestrationRun, **kwargs: Any) -> Json:
        return self._http.post("/orchestration/v1/run", body=body, **kwargs)

    def continue_run(self, body: OrchestrationContinue, **kwargs: Any) -> Json:
        return self._http.post("/orchestration/v1/run/continue", body=body, **kwargs)
